### ISSUE TITLE
Make lazy evaluation an explicit implementation-defined feature

### ISSUE THREAD

**Issue by ndw** [xproc/3.0-specification#1113]
In 17.2, we observe that some implementations may evaluate expressions lazily and others eagerly and this may have effect whether or not a test raises a dynamic error. But we don't call this out as an explicit, implementation-defined feature.

I think we should. Users should know what behavior a processor uses.

**Comment by ndw** [comment]
I've lost confidence in this idea. If we make lazy evaluation into a feature, it would be reasonable for users to expect some level of interoperability with the feature. But this is partly about optimizations performed by the implementation and those are just hard to make interoperable. There's all sorts of scope for implementors to do lazy evaluation in some places, under some circumstances, and not in others. Maybe an implementation wants to do eager evaluation of statically computable values and lazy evaluation of runtime values. Is that wrong? Does that choice have to be explicitly described by the specification?

Maybe the spec says enough as it is?

**Comment by gimsieke** [comment]
I tend to agree, but maybe Achim as an implementer has a more well-founded opinion.

**Comment by gimsieke** [comment]
or is it “better-founded”?

**Comment by xml-project** [comment]
@ndw says
> Maybe the spec says enough as it is?
I think it would be very hard to says if a given expression is evaluated lazy or not. This might depend on the context (surrounding steps - previous and/or following) and even (if you think about document projection) on source documents.
My proposal would be to leave the specification as it is for now. If we run into troubles reported by users, we may update 17.2 later.
Independ of whether an expression/step is lazy evaluated or not, if pipeline authors aim on interoperable pipelines they should catch any possible error. So the question is not, is an error actually raised or not, but **could** a compliant processor raise an error. Looking the problem this way, the spec says all, a pipeline author needs to know.
Does this make sense?

**Comment by ndw** [comment]
As we discussed on the editor's call on 2 January 2025, it's unclear how we could usefully improve the discussion.
